app/src/org/commcare/android/framework/SessionAwareCommCareActivity.java

```diff
@@ -1,11 +1,30 @@
 package org.commcare.android.framework;
 
+import android.os.Bundle;
+
+import org.commcare.android.util.SessionUnavailableException;
+
 /**
  * Manage redirection to login screen when session expiration occurs.
  *
  * @author Phillip Mates (pmates@dimagi.com)
  */
 public abstract class SessionAwareCommCareActivity<R> extends CommCareActivity<R> {
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        try {
+            onCreateAware(savedInstanceState);
+        } catch (SessionUnavailableException e) {
+            SessionActivityRegistration.redirectToLogin(this);
+        }
```

Comment:
@ctsims @amstone326 here is the commit with my approach to this issue: https://github.com/dimagi/commcare-odk/commit/973f70d9730e28a77a47e0eb374f8e2cb9b94013

Main idea is to fail soft with certain isPinSet checks when the result only effects the UI, which won't even be shown because the user will be redirected to the login screen before it is shown.

I'm not in love with the solution but it localized exception handling instead of integrating it into our session framework, which I think is important.